### PR TITLE
`Programming exercises`: Fix missing build plan during import on Jenkins setups

### LINKED ISSUE #7056: `Programming exercises`: Missing build plan when importing programming exercises
### Describe the bug

After importing a programming exercise, the build plan is missing. No inputs can be made, and the wheel in the top bar keeps on spinning.
The re-create build plan option also doesn't work.

### To Reproduce

1. Import a programming exercise from instance (meaning from another course, for example)
2. Click on 'Edit Build Plan'
3. The editor is empty and no inputs can be made

### Expected behavior

I expect the import to also include the build plan.

### Screenshots

bpe

### Which version of Artemis are you seeing the problem on?

6.3.8

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by chrisknedl**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Closes #7056 

### Description

In order to easily access the exercises' build plans more easily in the `JenkinsBuildPlanService`, the `copyBuildPlan` method in the `ContinuousIntegrationService` interface is changed to pass the complete programming exercise objects instead of only the project keys. The keys can still be easily retrieved from the exercises by calling the `getProjectKey` method. In addition, a new method is added to the `BuildPlanRepository`. That one is used to actually copy the new build plan from source to target for the Jenkins setups.

Fixed import process:
- Connect the build plan in the Artemis database additionally to the new exercise.
- In Jenkins: make sure the URL from which the build plan is downloaded (Artemis REST-API) is updated to link to the new programming exercise ID and secret token.
  - Otherwise the build plan of the original exercise would still be used.
  - This is done directly as part of the copy build plan process, i.e., Artemis fetches the old build plan, updates the necessary URL, and copies is back into the correct directory in Jenkins.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise
- test server with GitLab/Jenkins setup

1. Log in to Artemis
2. Navigate to Course Management
3. Click `Exercises`
4. Click `Import Programming Exercise` and make sure `From Instance` is selected.
5. Click the import button next to the existing exercise.
6. Fill in the required fields and save the exercise.
7. Click on the `Edit Build Plan` button.
8. Make sure that the build plan is present and is the same as the one from the source exercise.

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ProgrammingExerciseImportBasicService.java | 100% | ✅                           |
| ProgrammingExerciseImportService.java | 98% | ✅                           |
| JenkinsBuildPlanService | 94% | ✅                           |

### CODE REVIEW

**Comment by chrisknedl** [comment]
Currently still kind of flaky, as hibernate only sometimes decides to save the new build plan access secret.

**Review by b-fein — APPROVED** [review]
Code looks good.
Also tested the import on a Uni Passau testserver with and without recreation of build plans. Works as expected.

**Review by reschandreas — DISMISSED** [review]
tested locally, works as expected, thanks for fixing this!

**Comment by chrisknedl** [comment]
@reschandreas Thanks a lot for testing!

**Review by Gusti2010 — APPROVED** [review]
Tested on TS2 and works as expected. Build plan is the same in source and imported exercise. 
However the solution Solution Participation is "No corresponding result available" in the imported exercise. "View last result shows" 100% tho. I reproduced this with java and python exercises.

**Review by JanaNF — APPROVED** [review]
Manual tested on legacy_ts2. Worked as expected.

**Review by dfuchss — APPROVED** [review]
Code approval :)

**Review by tobias-lippert — DISMISSED** [review]
Code

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/build_plan/JenkinsBuildPlanService.java [hunk 1]
```diff
@@ -191,6 +196,34 @@ public void updateBuildPlanRepositories(String buildProjectKey, String buildPlan
             log.error("Pipeline Script not found", e);
         }
 
+        postBuildPlanConfigChange(buildPlanKey, buildProjectKey, jobConfig);
+    }
+
+    /**
+     * Replaces the old build plan URL with the new one, and posts the change to the Jenkins job config of the new exercise.
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/build_plan/JenkinsBuildPlanService.java:203]:
```suggestion
     * Replaces the old build plan URL with a new one containing an updated exercise and access token.
```

**Review by tobias-lippert — APPROVED** [review]
Reapprove

**Review by b-fein — APPROVED** [review]
Reapprove code+manual test after changes.

**Review by dfuchss — APPROVED** [review]
code lgtm

**Comment by MaximilianJG** [comment]
Tested on Testserver 2: Importing an exercise gives me the same build plan. LGTM!

**Review by egekurt123 — APPROVED** [review]
Manually tested on ts5, imported exercise has the same build plan as the original one

**Pull request merged**